inst/include/lpdens.hpp

```diff
@@ -5,231 +5,286 @@
-                                       bool inverse = false)
-    {
-        Eigen::VectorXd x_new = x;
-        if (!inverse) {
-            if (!std::isnan(xmin_) & !std::isnan(xmax_)) {  // two boundaries
-                x_new = (x.array() - xmin_ + 5e-3) / (xmax_ - xmin_ + 1e-2);
-                x_new = stats::qnorm(x_new);
-            } else if (!std::isnan(xmin_)) {                // left boundary
-                x_new = (1e-3 + x.array() - xmin_).log();
-            } else if (!std::isnan(xmax_)) {                // right boundary
-                x_new = (1e-3 + xmax_ - x.array()).log();
-            }
-        } else {
-            if (!std::isnan(xmin_) & !std::isnan(xmax_)) {  // two boundaries
-                x_new = stats::pnorm(x).array() + xmin_ - 5e-3;
-                x_new *=  (xmax_ - xmin_ + 1e-2);
-            } else if (!std::isnan(xmin_)) {                // left boundary
-                x_new = x.array().exp() + xmin_ - 1e-3;
-            } else if (!std::isnan(xmax_)) {                // right boundary
-                x_new = -(x.array().exp() - xmax_ - 1e-3);
-            }
-        }
-
-        return x_new;
-    }
-
+                                       bool inverse = false);
     Eigen::VectorXd boundary_correct(const Eigen::VectorXd& x,
-                                     const Eigen::VectorXd& fhat)
-    {
-        Eigen::VectorXd corr_term(fhat.size());
-        if (!std::isnan(xmin_) & !std::isnan(xmax_)) {  // two boundaries
-            corr_term = (x.array() - xmin_ + 5e-3) / (xmax_ - xmin_ + 1e-2);
-            corr_term = stats::dnorm(stats::qnorm(corr_term));
-            corr_term /= (xmax_ - xmin_ + 1e-2);
-            corr_term = 1 / corr_term.array().max(1e-4);
-        } else if (!std::isnan(xmin_)) {                // left boundary
-            corr_term = 1 / (1e-3 + x.array() - xmin_);
-        } else if (!std::isnan(xmax_)) {                // right boundary
-            corr_term = 1 / (1e-3 + xmax_ - x.array());
+                                     const Eigen::VectorXd& fhat);
+    Eigen::VectorXd construct_grid_points(const Eigen::VectorXd& x);
+    Eigen::VectorXd finalize_grid(Eigen::VectorXd& grid_points);
+    Eigen::VectorXd without_boundary_ext(const Eigen::VectorXd& grid_points);
+};
+
+//! constructor for fitting the density estimate.
+//! @param x vector of observations
+//! @param bw positive bandwidth parameter
+//! @param xmin lower bound for the support of the density, `NaN` means no
+//!   boundary.
+//! @param xmax upper bound for the support of the density, `NaN` means no
+//!   boundary.
+inline LPDens1d::LPDens1d(Eigen::VectorXd x,
+                          double bw,
+                          double xmin,
+                          double xmax) :
+    bw_(bw),
+    xmin_(xmin),
+    xmax_(xmax)
+{
+    if (bw <= 0)
+        throw std::runtime_error("bw must be positive.");
+
+    // remove NaNs from the data
+    x = tools::nan_omit(x);
```

Comment:
Aren't we doing this in R already?

Reply:
Yeah, right I added that later; will remove.